R/use_extendr.R

```diff
@@ -108,6 +128,44 @@ use_extendr <- function(path = ".", quiet = getOption("usethis.quiet", FALSE)) {
   return(invisible(TRUE))
 }
 
+#' Checks if provided name is a valid Rust name (identifier)
+#'
+#' @param name \[ character(n) \] Names to test.
+#' @return \[ logical(n) \] `TRUE` if the name is valid, otherwise `FALSE`.
+#' @noRd
+is_valid_rust_name <- function(name) {
+  # We require the name starts with a letter,
+  # ends with a letter or digit,
+  # and contains only alphanumeric ASCII chars, `-` or `_`.
+  stringi::stri_detect_regex(name, "^[A-z][\\w-]*[A-z0-9]$")
+}
+
+#' Convert R package name into equivalent valid Rust name.
+#'
+#' @param r_name \[ character(n) \] R names to convert.
+#' @return \[ character(n) \] Equivalent Rust name (if exists), otherwise `NA`.
+#' @noRd
+as_valid_rust_name <- function(r_name) {
+  rust_name <- stringi::stri_replace_all_fixed(r_name, ".", "_")
+  throw_if_invalid_rust_name(rust_name)
+  rust_name
+}
+
+#' Verifies if a function argument is a valid Rust name.
+#'
+#' @param name \[ string \] Tested caller function argument.
+#' @param call \[ env \] Environment of the caller, passed to `ui_throw()`.
+#' @noRd
+throw_if_invalid_rust_name <- function(name, call = caller_env()) {
+  quo <- enquo(name)
+  if (!is.character(name) || length(name) != 1 || !is_valid_rust_name(name)) {
```

Comment:
```suggestion
  if (length(name) != 1 || !is.character(name) || !is_valid_rust_name(name)) {
```

`is.character()` might return length >1 vector when `name` is length >1.

Reply:
I hate base R conventions so much sometimes. We should probably explicitly reference `{vctrs}` in the future and use it for checks.

Reply:
Ah, I just remembered we can simply use `rlang::is_scalar_character()`.